app/Main.hs

```diff
@@ -38,109 +48,120 @@ defaultProject =
+
+parseOther :: Parser OtherOptions
+parseOther = OtherOptions
+  <$> switch (long "no-core" <> help "Don't load Core.carp")
+  <*> switch (long "no-profile" <> help "Don't load profile.carp")
+  <*> switch (long "log-memory" <> help "Log memory allocations")
+  <*> switch (long "optimize" <> help "Optimized build")
+  <*> switch (long "generate-only" <> help "Stop after generating the C code")
+  <*> optional (strOption (long "prompt" <> help "Set REPL prompt"))
 
--- | Parse the arguments sent to the compiler from the terminal.
--- | TODO: Switch to 'cmdargs' library for parsing these!
-parseArgs :: [String] -> ([FilePath], ExecutionMode, [OtherOptions])
-parseArgs args = parseArgsInternal [] Repl [] args
-  where parseArgsInternal filesToLoad execMode otherOptions [] =
-          (filesToLoad, execMode, otherOptions)
-        parseArgsInternal filesToLoad execMode otherOptions (arg:restArgs) =
-          case arg of
-            "-b" -> parseArgsInternal filesToLoad Build otherOptions restArgs
-            "-x" -> parseArgsInternal filesToLoad BuildAndRun otherOptions restArgs
-            "-i" -> parseArgsInternal filesToLoad (Install (head restArgs)) otherOptions (tail restArgs)
-            "--check" -> parseArgsInternal filesToLoad Check otherOptions restArgs
-            "--no-core" -> parseArgsInternal filesToLoad execMode (NoCore : otherOptions) restArgs
-            "--no-profile" -> parseArgsInternal filesToLoad execMode (NoProfile : otherOptions) restArgs
-            "--log-memory" -> parseArgsInternal filesToLoad execMode (LogMemory : otherOptions) restArgs
-            "--optimize" -> parseArgsInternal filesToLoad execMode (Optimize : otherOptions) restArgs
-            "--generate-only" -> parseArgsInternal filesToLoad execMode (GenerateOnly : otherOptions) restArgs
-            "--prompt" -> case restArgs of
-                             newPrompt : restRestArgs ->
-                               parseArgsInternal filesToLoad execMode (SetPrompt newPrompt : otherOptions) restRestArgs
-                             _ ->
-                               error "No prompt given after --prompt"
-            file -> parseArgsInternal (filesToLoad ++ [file]) execMode otherOptions restArgs
+parseExecMode :: Parser ExecutionMode
+parseExecMode =
+  flag' Check (long "check" <> help "Check project")
+  <|> flag' Build (short 'b' <> help "Build project")
+  <|> flag' BuildAndRun (short 'x' <> help "Build an run project")
+  <|> Install <$> strOption (short 'i' <> help "Install built product")
+  <|> pure Repl
 
-setCustomPromptFromOptions :: Project -> [OtherOptions] -> Project
-setCustomPromptFromOptions project (o:os) =
-  case o of
-    SetPrompt newPrompt -> setCustomPromptFromOptions (project { projectPrompt = newPrompt }) os
-    _ -> setCustomPromptFromOptions project os
-setCustomPromptFromOptions project _ =
-  project
+parseFiles :: Parser [FilePath]
+parseFiles = some (argument str (metavar "FILES...")) <|> pure []
```

Comment:
I don't understand this piece of the code, what is the metavar thingy?

Reply:
The metavar is used in `--help`. If you don't set it, it will say `ARG` instead IIRC.

Reply:
The `<|> pure []` is to allow an empty argument list. Maybe there's a better way. You could use `optional` instead to return a `Maybe [String]` instead. On a second thought, this should be unified with `optPreload/Postload`, let me change it.

Reply:
`many` was the right answer. I tried it at some point and it didn't work (there was probably some ambiguity or an error somewhere else). It seems to work correctly now.